source/agora/test/QuorumShuffle.d

```diff
@@ -0,0 +1,469 @@
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.QuorumShuffle;
+
+version (unittest):
+
+import agora.api.Validator;
+import agora.common.Amount;
+import agora.common.Config;
+import agora.consensus.data.ConsensusParams;
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.consensus.data.Block;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import agora.consensus.data.genesis.Test;
+import agora.node.FullNode;
+import agora.test.Base;
+
+import std.algorithm;
+import std.format;
+import std.range;
+
+import core.thread;
+import core.time;
+
+/// With a validator cycle of 10 and shuffle cycle of 3 we should expect
+/// quorums being shuffled at these block heights:
+/// 3, 6, 9, 10 (enrollment change), 13 (next shuffle cycle)
+unittest
+{
+    import agora.common.Types;
+    TestConf conf = {
+        validators : 6,
+        max_listeners : 7,
+        extra_blocks : 2,  // 1 short of shuffle cycle
+        validator_cycle : 10,
+        max_quorum_nodes : 4,  // makes it easier to test shuffle cycling
+        quorum_shuffle_cycle : 3
+    };
+    auto network = makeTestNetwork(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    //scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 2, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 2)));
+
+    enum quorums_1 = [
+        // 0
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 1
+        QuorumConfig(4, [
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.E.address]),
+
+        // 2
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.B.address,
+            WK.Keys.C.address]),
+
+        // 3
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 4
+        QuorumConfig(4, [
+            WK.Keys.B.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address,
+            WK.Keys.E.address]),
+
+        // 5
+        QuorumConfig(4, [
+            WK.Keys.D.address,
+            WK.Keys.F.address,
+            WK.Keys.A.address,
+            WK.Keys.C.address]),
+    ];
+
+    version (none)
+    foreach (idx, node; nodes.enumerate)
+        writefln("Node %s: %s\n", idx, node.getQuorumConfig);
```

Comment:
Leftover ?

Reply:
It's inside of `version (none)`, I left it in purposefully for when we change the quorum generator settings. Then you can uncomment `version (none)` and easily modify the test when the results are expected to change.

Reply:
Should I name it something like `version (PrintQuorums)`? I guess that could prevent needing to recompile.

Reply:
Prior art: https://github.com/Geod24/localrest/blob/37460b237c3bf14544dcd46d8d673bbe36d2d4db/source/geod24/LocalRest.d#L1437-L1438

Reply:
`version (none)` works for me, was just thinking you forgot them. I'm happy to keep them.